### PR TITLE
Svelte: Support TypeScript and preprocessors

### LINKED ISSUE #12754: Official svelte typescript support
As Svelte [does now officially support the typescript language](https://svelte.dev/blog/svelte-and-typescript),
I decided to do the migration by following the linked official guide.

Unfortunately, this creates a crash issue on storybook launch (`npm run storybook`):

```
ERROR in ./src/components/Copy.svelte
Module build failed (from ./node_modules/svelte-loader/index.js):
Error: ParseError: Unexpected token (4:20)
2:   import { createEventDispatcher } from 'svelte';
3: 
4:   export let content: string;
                       ^
5:   export let showContent: boolean = true;
6:   export let title: string = 'Copy content';
    at /home/sullivan/p/gitlab.com/nexylan/svelty/node_modules/svelte-loader/index.js:181:12
 @ ./src/components/Copy.stories.js 65:0-33 69:13-17 73:15-19 98:15-19
 @ ./src/components sync ^\.(?:(?:^|\/|(?:(?:(?!(?:^|\/)\.).)*?)\/)(?!\.)(?=.)[^/]*?\.stories\.js)$
 @ ./.storybook/generated-stories-entry.js
 @ multi ./node_modules/@storybook/core/dist/server/common/polyfills.js ./node_modules/@storybook/core/dist/server/preview/globals.js ./.storybook/storybook-init-framework-entry.js ./node_modules/@storybook/addon-docs/dist/frameworks/common/config.js-generated-other-entry.js ./node_modules/@storybook/addon-actions/dist/preset/addDecorator.js-generated-other-entry.js ./node_modules/@storybook/addon-actions/dist/preset/addArgs.js-generated-other-entry.js ./node_modules/@storybook/addon-backgrounds/dist/preset/defaultParameters.js-generated-other-entry.js ./node_modules/@storybook/addon-a11y/dist/a11yRunner.js-generated-other-entry.js ./node_modules/@storybook/addon-a11y/dist/a11yHighlight.js-generated-other-entry.js ./.storybook/preview.js-generated-config-entry.js ./.storybook/generated-stories-entry.js (webpack)-hot-middleware/client.js?reload=true&quiet=false&noInfo=undefined
```

It looks like Storybook is not going to interpret typescript as it should.

Maybe I have to change some configuration for Storybook, but I can't found anything about this on the documentation and the issues.

Does Storybook support Svelte with Typescript?

You may test the issue by cloning this branch: https://gitlab.com/nexylan/svelty/-/merge_requests/67

### LINKED ISSUE #13902: Svelte Typescript docgen parsing error
**Describe the bug**
The same issue is mentioned in https://github.com/storybookjs/storybook/issues/12754, thought would be good to create a separated issue to follow up fix. Apparently, TS inside `<script>` breaks `svelte-docgen-loader` parsing and generate incomplete props table (missing `Description` and `Default` values).

**To Reproduce**

1. create a Svelte app from template `npx create-snowpack-app my-app --template app-template-sstt` 
2. `cd my-app`
3. Init Storybook without installing packages (due to peer dependency conflict) `npx -p @storybook/cli sb init -s`
4. `npm i --legacy-peer-deps` to install Storybook packages
5. Add type `let test: string;`  in any example component (e.g. `Button`)

**Expected behavior**
Screen Shot 2021-02-14 at 12 21 47 pm

**Screenshots**
Screen Shot 2021-02-14 at 11 54 20 am
Screen Shot 2021-02-14 at 12 01 14 pm

**Code snippets**
```typescript
<script lang="ts">
  import { onMount } from "svelte";

  // espree parser cannot recognise ":"
  let cardRef: HTMLElement;
  let containerRef: HTMLElement;
</script>
```

**System**
Please paste the results of `npx sb@next info` here.

**Additional context**

### PR DESCRIPTION
**Pull request by j3rem1e**
Issue: #12754

## What I did

Added support for preprocessor in svelte, with a `svelteOptions.preprocessor` properties. the preprocessors are used in `svelte-loader` and in `svelte-docgen-loader`. Additionnaly, `svelteOptions.loader` are passed to the `svelte-loader` plugin.

`svelte-kitchen-sink` has been configured to use typescript, and the button component is now a typescript component.
The test in svelte-kitchen-sink used a custom jest transformation for the svelte component. however, this script was not supporting preprocessors. I updated the example to use a "real" jest transform (svelte-jester) and updated the tests. This is just local to svelte-kitchen-sink, there is no impact to the users, who probably already use a jest transform.

## How to test

- Is this testable with Jest or Chromatic screenshots? Done
- Does this need a new example in the kitchen sink apps? Done
- Does this need an update to the documentation? No

### CODE REVIEW

File: addons/docs/package.json [hunk 1]
```diff
@@ -130,6 +132,7 @@
     "babel-loader": "^8.0.0",
     "react": "^16.8.0 || ^17.0.0",
     "react-dom": "^16.8.0 || ^17.0.0",
+    "svelte": "^3.31.2",
```
Comment (shilman) [inline addons/docs/package.json:135]: needs to be an optional peer dependency also. hopefully we can refactor this soon cc @phated

File: addons/docs/src/frameworks/svelte/svelte-docgen-loader.ts [hunk 2]
```diff
@@ -64,7 +83,9 @@ export default async function svelteDocgen(source: string) {
               ${componentName}.__docgen = ${JSON.stringify(componentDoc)};
               `;
   } catch (error) {
-    console.error(error);
+    if (logDocgen) {
+      console.error(error);
```
Comment (shilman) [inline addons/docs/src/frameworks/svelte/svelte-docgen-loader.ts:87]: let's use `@storybook/node-logger` here for consistency

File: examples/svelte-kitchen-sink/svelte.config.js [hunk 3]
```diff
@@ -0,0 +1,5 @@
+const sveltePreprocess = require('svelte-preprocess');
+
+module.exports = {
+  preprocess: sveltePreprocess(),
```
Comment (shilman) [inline examples/svelte-kitchen-sink/svelte.config.js:4]: what's the purpose of this `preprocess` vs. the call in `main.js/svelteOptions`?

**Review by shilman — CHANGES_REQUESTED** [review]
Looking great--this is fantastic! I have a couple minor requests and there's a piece I want to understand better. Aside from that this is ready to go!

File: examples/svelte-kitchen-sink/svelte.config.js (hunk 3 above)
Comment (j3rem1e) [reply examples/svelte-kitchen-sink/svelte.config.js:4]: `svelte.config.js` is an unofficial configuration file used by some tools to centralize the configuration of the svelte compiler.

It's not used by svelte itself, but by the vscode extension, or here by `svelte-jester`.

This file is undocumented/unofficial, so it didn't try to integrate it into storybook. it's just here in the svelte-kitchen-sink to make svelte-jester work.

**Review by shilman — APPROVED** [review]
Thanks for the clarification @j3rem1e 🙏  Looking good -- will merge/release once CI passes

**Pull request merged**